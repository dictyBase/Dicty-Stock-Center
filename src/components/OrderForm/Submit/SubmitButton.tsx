import React from "react"
import { useFormikContext } from "formik"
import Button from "@material-ui/core/Button"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { useMutation, useLazyQuery } from "@apollo/react-hooks"
import { useHistory } from "react-router-dom"
import { useCartStore } from "components/ShoppingCart/CartStore"
import useCartItems from "hooks/useCartItems"
import { CREATE_ORDER, CREATE_USER, UPDATE_USER } from "graphql/mutations"
import { GET_USER_BY_EMAIL } from "graphql/queries"
import useStyles from "../formStyles"
import { FormikValues } from "../utils/initialValues"
import { CartItem } from "../types"

const getIDs = (items: Array<CartItem>) =>
  items.map((item: CartItem) => item.id)

/**
 * getUserVariables generates a variables object that is passed with
 * create or update user mutations.
 */
const getUserVariables = (values: FormikValues, id?: string) => {
  const variablesObj: any = {
    variables: {
      input: {
        first_name: values.firstName,
        last_name: values.lastName,
        email: values.email,
        organization: values.organization,
        group_name: values.lab,
        first_address: values.address1,
        second_address: values.address2,
        city: values.city,
        state: values.state,
        zipcode: values.zip,
        country: values.country,
        phone: values.phone,
        is_active: true,
      },
    },
  }
  // if there's an ID, that means we need to send an update mutation;
  // add ID to variables and remove email (since this cannot be updated)
  if (id) {
    variablesObj.variables.id = id
    delete variablesObj.variables.input.email
  }
  return variablesObj
}

/**
 * getOrderVariables generates a variables object that is sent with the
 * create order mutation.
 */
const getOrderVariables = (
  values: FormikValues,
  addedItems: Array<CartItem>,
) => ({
  variables: {
    input: {
      courier: values.shippingAccount,
      courier_account: values.shippingAccountNumber,
      comments: values.comments,
      payment: values.paymentMethod,
      purchase_order_num: values.purchaseOrderNum,
      status: "IN_PREPARATION",
      consumer: values.email,
      payer: values.payerEmail,
      purchaser: values.email,
      items: getIDs(addedItems),
    },
  },
})

/**
 * SubmitButton is the button used to submit the order. It
 * appears on the last page of the order form, and it contains the
 * necessary logic for GraphQL queries and mutations.
 */

const SubmitButton = ({ setSubmitError }: { setSubmitError: Function }) => {
  const { isSubmitting, values, submitForm } = useFormikContext<any>()
  const [{ addedItems }] = useCartStore()
  const { emptyCart } = useCartItems(addedItems)
  const history = useHistory()
  const [createOrder] = useMutation(CREATE_ORDER, {
    onCompleted: () => {
      setSubmitError(false)
      submitForm()
      history.push("/order/submitted")
      emptyCart()
    },
    onError: () => {
      setSubmitError(true)
    },
  })
  const [createUser] = useMutation(CREATE_USER, {
    onCompleted: () => {
      setSubmitError(false)
      createOrder(getOrderVariables(values, addedItems))
    },
    onError: () => {
      setSubmitError(true)
    },
  })
  const [updateUser] = useMutation(UPDATE_USER, {
    onCompleted: () => {
      setSubmitError(false)
      createOrder(getOrderVariables(values, addedItems))
    },
    onError: () => {
      setSubmitError(true)
    },
  })
  const [getUserByEmail] = useLazyQuery(GET_USER_BY_EMAIL, {
    onCompleted: (data) => {
      updateUser(getUserVariables(values, data.userByEmail.id))
      setSubmitError(false)
    },
    onError: (error: any) => {
      if (
        error.graphQLErrors.length > 0 &&
        error.graphQLErrors[0].extensions.code === "NotFound"
      ) {
        setSubmitError(false)
        createUser(getUserVariables(values))
        return
      }
      setSubmitError(true)
    },
  })
  const classes = useStyles()

  const handleSubmit = () => {
    getUserByEmail({
      variables: { email: values.email },
    })
  }

  return (
    <Button
      type="submit"
      size="large"
      className={classes.submitBtn}
      disabled={isSubmitting}
      onClick={handleSubmit}>
      Submit Order &nbsp;
      <FontAwesomeIcon icon="check-circle" />
    </Button>
  )
}

export default SubmitButton
