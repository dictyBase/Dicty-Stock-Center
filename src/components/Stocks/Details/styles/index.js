import { makeStyles } from "@material-ui/core/styles"
import { green } from "@material-ui/core/colors"

const useStyles = makeStyles(({ palette, spacing }) => ({
  root: {
    width: "90%",
    margin: "auto",
  },
  header: {
    textAlign: "center",
  },
  backButton: {
    marginTop: "10px",
    textAlign: "left",
    color: "#5f6368",
    "&:hover": {
      backgroundColor: "transparent",
      color: "inherit",
    },
  },
  cardHeader: {
    color: "#002f5e",
  },
  details: {
    "&:nth-child(odd)": {
      backgroundColor: "#F4F6F8",
    },
    "@media (max-width: 420px)": {
      display: "block !important",
    },
  },
  listItem: {
    "@media (max-width: 420px)": {
      marginBottom: "10px",
    },
  },
  list: {
    width: "100%",
    padding: "0px !important",
  },
  availabilityCard: {
    paddingTop: "5px",
    textAlign: "center",
  },
  cartBtnRow: {
    padding: "16px",
  },
  unavailableText: {
    padding: "16px",
    textAlign: "justify",
  },
  moreStrainsCard: {
    marginTop: "16px",
    padding: "5px",
    textAlign: "center",
  },
  options: {
    paddingBottom: "8px",
    display: "flex",
    justifyContent: "center",
  },
  secondaryText: {
    paddingTop: "8px",
  },
  link: {
    textDecoration: "none",
    color: palette.primary.main,
  },
  checkoutBtn: {
    borderStyle: "none !important",
    borderRadius: "0px !important",
    fontSize: "0.9rem !important",
  },
  arrowIcon: {
    marginRight: "8px",
  },
  id: {
    display: "flex",
    justifyContent: "center",
  },
  snackbar: {
    width: "300px",
    backgroundColor: green[600],
    fontSize: "0.8rem",
    display: "flex",
    alignItems: "center",
    paddingTop: "0px",
    paddingBottom: "0px",
  },
  checkIcon: {
    fontSize: "0.8rem",
    marginRight: spacing(1),
  },
  closeIcon: {
    fontSize: "0.9rem",
  },
  unavailableIcon: {
    marginRight: spacing(2),
  },
  copyIcon: {
    fontSize: "0.8rem",
    "&:hover": {
      backgroundColor: "transparent",
      color: "rgba(0, 0, 0, 0.87)",
    },
  },
  detailsItem: {
    fontWeight: 420,
  },
  relatedItem: {
    display: "flex",
    justifyContent: "center !important",
  },
  addToCartBtn: {
    backgroundColor: green[600],
    color: "#fff",
    paddingLeft: "16px",
    paddingRight: "16px",
    "&:hover": {
      backgroundColor: green[700],
      color: "#fff",
    },
  },
  removeBtn: {
    marginTop: "5px",
  },
  quantity: {
    display: "flex",
    justifyContent: "space-evenly",
    alignItems: "center",
    padding: "16px",
  },
  maxItems: {
    color: palette.secondary.main,
  },
}))

export default useStyles
