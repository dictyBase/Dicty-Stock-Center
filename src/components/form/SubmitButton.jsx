import React, { Component } from 'react'
import PropTypes from 'prop-types'
import 'styles/core.scss'

export default class SubmitButton extends Component {
    displayName = 'form submit button';

    static propTypes = {
        submitting: PropTypes.bool,
        name: PropTypes.string,
        icon: PropTypes.string
    }

    render() {
        const { submitting, name, icon } = this.props
        return (
            <button type="submit"
                className="btn btn-primary btn-lg btn-block"
                disabled={ submitting }> { name }
                { submitting ? <i className="fa fa-spinner fa-pulse fa-fw margin-bottom"></i>
                    : <i className={ icon && icon } aria-hidden="true"></i>
                }
            </button>
        )
    }
}
