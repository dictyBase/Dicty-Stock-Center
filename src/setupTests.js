import { configure } from "enzyme"
import Adapter from "enzyme-adapter-react-16"
import "jest-enzyme"

global.console = {
  log: jest.fn(),
  error: jest.fn(),
  warn: jest.fn(),
}

configure({ adapter: new Adapter() })
