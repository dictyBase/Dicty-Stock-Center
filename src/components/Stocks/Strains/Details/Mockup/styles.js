import { makeStyles } from "@material-ui/core/styles"

const useStyles = makeStyles({
  root: {
    width: "80%",
    margin: "auto",
  },
  header: {
    textAlign: "center",
  },
  leftCard: {},
  details: {
    "&:hover": {
      boxShadow:
        "inset 1px 0 0 #dadce0,inset -1px 0 0 #dadce0,0 1px 2px 0 rgba(60,64,67,.3),0 1px 3px 1px rgba(60,64,67,.15)",
      zIndex: 1,
    },
    "&:nth-child(odd)": {
      backgroundColor: "#eeeeee",
    },
  },
  listTitle: {},
  listContent: {},
  rightCard: {
    padding: "5px",
    textAlign: "center",
  },
  quantity: {
    display: "inline-flex",
    justifyContent: "center",
  },
  textField: {},
})

export default useStyles
