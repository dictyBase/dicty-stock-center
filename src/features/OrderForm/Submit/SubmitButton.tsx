import React from "react"
import { makeStyles, Theme } from "@material-ui/core/styles"
import Button from "@material-ui/core/Button"
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { useMutation, useQuery } from "@apollo/client"
import { useHistory } from "react-router-dom"
import { useCartStore } from "features/ShoppingCart/CartStore"
import useCartItems from "common/hooks/useCartItems"
import {
  CREATE_ORDER,
  CREATE_USER,
  UPDATE_USER,
} from "common/graphql/mutations"
import { GET_USER_BY_EMAIL } from "common/graphql/queries/user"
import { FormikValues } from "../utils/initialValues"
import { CartItem } from "common/types"

/**
 * getIDs creates a new array of just stock IDs
 */
const getIDs = (items: Array<CartItem>) =>
  items.map((item: CartItem) => item.id)

const getConsumerVariables = (formData: FormikValues) => ({
  first_name: formData.firstName,
  last_name: formData.lastName,
  email: formData.email,
  organization: formData.organization,
  group_name: formData.lab,
  first_address: formData.address1,
  second_address: formData.address2,
  city: formData.city,
  state: formData.state,
  zipcode: formData.zip,
  country: formData.country,
  phone: formData.phone,
  is_active: true,
})

const getPayerVariables = (formData: FormikValues) => ({
  first_name: formData.payerFirstName,
  last_name: formData.payerLastName,
  email: formData.payerEmail,
  organization: formData.payerOrganization,
  group_name: formData.payerLab,
  first_address: formData.payerAddress1,
  second_address: formData.payerAddress2,
  city: formData.payerCity,
  state: formData.payerState,
  zipcode: formData.payerZip,
  country: formData.payerCountry,
  phone: formData.payerPhone,
  is_active: true,
})

/**
 * getUserVariables generates a variables object that is passed with
 * create or update user mutations.
 */
const getUserVariables = (
  formData: FormikValues,
  userType: string,
  id?: string,
) => {
  const inputVals =
    userType === "consumer"
      ? getConsumerVariables(formData)
      : getPayerVariables(formData)
  const variablesObj: any = {
    variables: {
      input: inputVals,
    },
  }
  // if there's an ID, that means we need to send an update mutation;
  // add ID to variables and remove email (since this cannot be updated)
  if (id) {
    variablesObj.variables.id = id
    delete variablesObj.variables.input.email
  }
  return variablesObj
}

/**
 * getOrderVariables generates a variables object that is sent with the
 * create order mutation.
 */
const getOrderVariables = (
  formData: FormikValues,
  addedItems: Array<CartItem>,
) => ({
  variables: {
    input: {
      courier: formData.shippingAccount,
      courier_account: formData.shippingAccountNumber,
      comments: formData.comments,
      payment: formData.paymentMethod,
      purchase_order_num: formData.purchaseOrderNum,
      status: "IN_PREPARATION",
      consumer: formData.email,
      payer: formData.payerEmail,
      purchaser: formData.email,
      items: getIDs(addedItems),
    },
  },
})

/**
 * updateOrCreateUser attempts to find a user from our database via the consumer's
 * email address. If successful, it then updates the user with the current formData.
 * Otherwise, it adds a new user to our database.
 */
const updateOrCreateUser = async (
  refetch: Function,
  formData: FormikValues,
  updateUser: Function,
  createUser: Function,
  setSubmitError: Function,
  userType: string,
) => {
  const userEmail =
    userType === "consumer" ? formData.email : formData.payerEmail
  try {
    const user = await refetch({
      email: userEmail,
    })
    if (user.data.userByEmail) {
      const updatedUser = await updateUser(
        getUserVariables(formData, userType, user.data.userByEmail.id),
      )
      return updatedUser
    }
  } catch (error) {
    const notFound = error.graphQLErrors[0].extensions.code === "NotFound"
    if (notFound) {
      const createdUser = await createUser(getUserVariables(formData, userType))
      return createdUser
    }
    setSubmitError(error)
    return Promise.reject(error)
  }
}

const useStyles = makeStyles((theme: Theme) => ({
  button: {
    minWidth: "200px",
  },
}))

type Props = {
  /** Full object of form data (shipping and payment) */
  formData: FormikValues
  /** Function to set a submit error (bool) */
  setSubmitError: Function
}

/**
 * SubmitButton is the button used to submit the order. It
 * appears on the last page of the order form, and it contains the
 * necessary logic for GraphQL queries and mutations.
 */

const SubmitButton = ({ formData, setSubmitError }: Props) => {
  const {
    state: { addedItems },
  } = useCartStore()
  const { emptyCart, getCartTotal } = useCartItems()
  const history = useHistory()
  const [createOrder] = useMutation(CREATE_ORDER)
  const [createUser] = useMutation(CREATE_USER)
  const [updateUser] = useMutation(UPDATE_USER)
  const { refetch } = useQuery(GET_USER_BY_EMAIL, {
    variables: { email: formData.email },
    skip: true, // skip initial fetch, we only want to fetch on button click
  })
  const classes = useStyles()

  const handleSubmit = async () => {
    try {
      // update or create consumer
      await updateOrCreateUser(
        refetch,
        formData,
        updateUser,
        createUser,
        setSubmitError,
        "consumer",
      )
      // update or create payer
      await updateOrCreateUser(
        refetch,
        formData,
        updateUser,
        createUser,
        setSubmitError,
        "payer",
      )
      const order = await createOrder(getOrderVariables(formData, addedItems))
      history.push("/order/submitted", {
        orderID: order.data.createOrder.id,
        formData,
        cartItems: addedItems,
        cartTotal: getCartTotal(addedItems),
      })
      emptyCart()
    } catch (error) {
      setSubmitError(true)
    }
  }

  return (
    <Button
      aria-label="Submit"
      type="submit"
      size="large"
      variant="contained"
      color="primary"
      className={classes.button}
      endIcon={<FontAwesomeIcon icon="check-circle" />}
      onClick={handleSubmit}>
      Submit Order
    </Button>
  )
}

export { getUserVariables, getIDs }
export default SubmitButton
