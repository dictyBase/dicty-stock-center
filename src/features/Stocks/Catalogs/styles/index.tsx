import { makeStyles, Theme } from "@material-ui/core/styles"
import { blue } from "@material-ui/core/colors"

const useStyles = makeStyles((theme: Theme) => ({
  appBar: {
    backgroundColor: blue[800],
  },
  catalogPaper: {
    height: 600,
    width: "100%",
  },
  listHeaders: {
    position: "sticky",
    top: 0,
  },
  row: {
    borderBottom: "1px solid rgba(224, 224, 224, 1)",
    "&:hover": {
      backgroundColor: "#eeeeee",
      boxShadow:
        "inset 1px 0 0 #dadce0,inset -1px 0 0 #dadce0,0 1px 2px 0 rgba(60,64,67,.3),0 1px 3px 1px rgba(60,64,67,.15)",
      zIndex: 1,
    },
    "@media (max-width: 1024px)": {
      "& p": {
        fontSize: "0.75rem !important",
      },
    },
  },
  item: {
    paddingRight: theme.spacing(1),
  },
  trashBtn: {
    color: theme.palette.error.main,
  },
}))

export default useStyles
