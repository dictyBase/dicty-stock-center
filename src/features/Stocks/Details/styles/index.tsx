import { makeStyles, Theme } from "@material-ui/core/styles"

// @ts-ignore <-- necessary to include the !important attributes
const useStyles = makeStyles(({ palette, spacing }: Theme) => ({
  header: {
    textAlign: "center",
  },
  cardHeader: {
    color: "#002f5e",
    paddingRight: "8px",
  },
  details: {
    "&:nth-child(odd)": {
      backgroundColor: "#F4F6F8",
    },
    "@media (max-width: 420px)": {
      display: "block !important",
    },
  },
  listItem: {
    "@media (max-width: 420px)": {
      marginBottom: "10px",
    },
  },
  list: {
    width: "100%",
    padding: "0px !important",
  },
  id: {
    display: "flex",
    justifyContent: "center",
  },
  snackbar: {
    width: "300px",
    backgroundColor: palette.secondary.main,
    fontSize: "0.8rem",
    display: "flex",
    alignItems: "center",
    paddingTop: "0px",
    paddingBottom: "0px",
  },
  checkIcon: {
    fontSize: "0.8rem",
    marginRight: spacing(1),
  },
  closeIcon: {
    fontSize: "0.9rem",
  },
  copyIcon: {
    fontSize: "0.8rem",
    "&:hover": {
      backgroundColor: "transparent",
      color: "rgba(0, 0, 0, 0.87)",
    },
  },
  detailsItem: {
    fontWeight: 420,
  },
  maxItems: {
    color: palette.error.main,
  },
  moreInfoText: {
    marginTop: "4px",
    fontSize: "0.75rem",
    color: "#000",
  },
  error: {
    color: palette.error.main,
  },
}))

export default useStyles
