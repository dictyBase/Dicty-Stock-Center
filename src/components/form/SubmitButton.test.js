import React from "react"
import "jest-styled-components"
import "../../setupTests"
import SubmitButton from "./SubmitButton"
import { shallow } from "enzyme"
import { PrimaryLargeButton } from "styles"

describe("form/SubmitButton", () => {
  let wrapper = shallow(
    <SubmitButton submitting={false} name={"Submit"} icon={"Submit"} />,
  )

  it("Should render as a <button>.", () => {
    expect(wrapper.is(PrimaryLargeButton)).toEqual(true)
  })
})
