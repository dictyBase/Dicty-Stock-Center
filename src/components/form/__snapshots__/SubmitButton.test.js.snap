// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matching a snapshot of SubmitButton 1`] = `
<button
  className="large block sc-gPEVay QPpEG"
  disabled={false}
  type="submit"
>
   
  Submit
  <span
    aria-hidden={true}
    className="fa fa-Submit"
  />
</button>
`;
